Add build support for arm64-linux

kaboing opened the pull request:
Adds support for arm64-linux.

Uses buildx to build multi-platform image.

Spawned by: #611

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://github.com/googleforgames/quilkin/blob/main/CONTRIBUTING.md 
   and developer guide https://github.com/googleforgames/quilkin/blob/main/build/README.md
2. Please label this pull request according to what type of issue you are addressing.
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/googleforgames/quilkin/blob/main/build/README.md#run-tests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup
> /kind documentation

/kind feature

**What this PR does / Why we need it**:
Adds support for arm64/linux. Builds multi-platform using buildx.  

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Closes #<issue number>`, or `Closes (paste link of issue)`.
-->
Closes #611 

**Special notes for your reviewer**:

kaboing committed:
Add build support for arm64-linux

Adds support for arm64-linux.

Uses buildx to build multi-platform image.

Spawned by: #611

kaboing commented:
/gcbrun

markmandel commented:
/gcbrun

markmandel commented:
Was `make build` working for you locally?

kaboing commented:
Yeah, it works on my machine ™️ . I build on an apple m2 pro with docker-desktop. 

I noticed that the error messages complains about building `ring`, but I don't know if that is relevant.

kaboing committed:
Add explicitly setting CC for aarch64-linux builds to fix cross-compilation issue with Ring

markmandel commented:
> Yeah, it works on my machine ™️ . I build on an apple m2 pro with docker-desktop.

Ah fun, so you would be building on ARM, and Cloud Build is building on Intel.

My suggestion would be to add the package `g++-aarch64-linux-gnu` to 

https://github.com/googleforgames/quilkin/blob/00345bec721076fa683ee01efce1db93341f1972/build/build-image/Dockerfile#L28-L31

Which should give the system what it needs.

kaboing committed:
Add package g++-aarch64-linux-gnu to quilkin-build

kaboing commented:
Gotcha. Let's test this version then!

markmandel commented:
/gcbrun

XAMPPRocky commented:
> Step #5 - "build": ERROR: multiple platforms feature is currently not supported for docker driver. Please switch to a different driver (eg. "docker buildx create --use")

kaboing committed:
Update multiarch build to use docker-container as build driver

kaboing commented:
Updated PR to use docker-container as build driver. However, according to docs, 
> Unlike docker driver, built images will not automatically appear in docker images and [build --load](https://docs.docker.com/reference/cli/docker/buildx/build/#load) needs to be used to achieve that.

Docs for `build --load`
> Shorthand for [--output=type=docker](https://docs.docker.com/reference/cli/docker/buildx/build/#docker). Will automatically load the single-platform build result to docker images.

=> does not seem to work for multi-arch images. So we perhaps need to restructure a bit to push directly, or configure some other reasonable output. Any thoughts on this?

markmandel commented:
> does not seem to work for multi-arch images

In my experience the multi arch image is stored temporarily in the build cache, so a subsequent `build` with `--push` will be almost immediate, because it won't rebuild.

markmandel commented:
Well it built! But the image could not be found after the fact.

kaboing committed:
Add --load to load image into Docker

kaboing commented:
Well, it's progress at least! And I see in the logs that it build for both architectures, so that's nice.
When trying `--load` locally (latest commit), the entire thing hanged on sending the resulting tarball, but let's at least try it in the gcb environment and see what happens.

kaboing committed:
Added --push to docker build command

Restructured test-agones to require build-image directly since it's already doing the pushing.

XAMPPRocky commented:
/gcbrun

XAMPPRocky committed:
Merge branch 'main' into build-for-arm64

XAMPPRocky commented:
So it successfully built, but the image uploaded to CI doesn't look multi arch to me.

```
docker buildx imagetools inspect us-docker.pkg.dev/quilkin/ci/quilkin:0.10.0-dev-0aa6b83

Name:      us-docker.pkg.dev/quilkin/ci/quilkin:0.10.0-dev-0aa6b83
MediaType: application/vnd.docker.distribution.manifest.v2+json
Digest:    sha256:5a56ba3b94663f46b4e0aa673735c662ad474c21dcb6a1ad908000f75b474985
```

kaboing commented:
No, I concur. One of the other manifests uploaded to the registry at the same time do have different entries for arm64 and amd64: 
https://console.cloud.google.com/artifacts/docker/quilkin/us/ci/quilkin/sha256:d9d655c2d817e95d85b56286cee74a633cb1efb6d32650fdcb145702454485aa;tab=manifest?inv=1&invt=AbmXyg

```
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.oci.image.index.v1+json",
   "manifests": [
      {
         "mediaType": "application/vnd.oci.image.manifest.v1+json",
         "digest": "sha256:3a22f7b558eb554c10c351db637d438803472ac59a6af5e5fd6c4ba357c3df3c",
         "size": 3895,
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.oci.image.manifest.v1+json",
         "digest": "sha256:4f2a123d6da1396fd987f71ae63f7597c180c76bde2c78199aec5d017f27cff1",
         "size": 3895,
         "platform": {
            "architecture": "arm64",
            "os": "linux"
         }
      },
      ...
}
```

I'm trying to figure out what more needs to be done. 

Just to make sure, I've tried to run the image created on arm64 and it does not work.

kaboing committed:
Explicitly creating docker manifest for multi-arch build

kaboing committed:
Explicitly create manifest and push

XAMPPRocky commented:
/gcbrun

XAMPPRocky commented:
Seeing the same result unfortunately

```
docker buildx imagetools inspect us-docker.pkg.dev/quilkin/ci/quilkin:0.10.0-dev-09bbe46
Name:      us-docker.pkg.dev/quilkin/ci/quilkin:0.10.0-dev-09bbe46
MediaType: application/vnd.docker.distribution.manifest.v2+json
Digest:    sha256:cf0e208da1ecd98500284826b1cfc20a8b04e3d8f857dc36d3d7a13073c550aa
```

kaboing commented:
Thanks for running the gcbruns for me on this PR, I'll investigate further where the problem is.

kaboing commented:
This digest should have the tag, it is the manifest list that points to the other manifests. Weird.
![CleanShot 2025-01-17 at 16 14 59@2x](https://github.com/user-attachments/assets/a5e305f9-d3d5-477a-a2b1-cde971e0e077)

![CleanShot 2025-01-17 at 16 16 43@2x](https://github.com/user-attachments/assets/f9645e2b-5733-40df-8964-8945a40d5e42)

kaboing committed:
Potential fix for docker manifest list tagging issue

Minor fix for quoting issue on package_version

XAMPPRocky commented:
/gcbrun

kaboing committed:
Reinstate buildx builds again

XAMPPRocky commented:
/gcbrun

maggie44 commented:
Is this how a working build and deploy ready for merging? Happy to jump in and take a look if not.

kaboing commented:
Yes, it's been tried and it ended up with the manifest list being wrongly tag. The explicit creation of manifest list was created in the hopes of fixing that problem, but we've ended up in the same situation. 

(also, usual caveats, I might have stumbled earlier and the proposed solution might work, but as far as I know, it's been tested)
